### PR TITLE
set transformer to evaluation mode

### LINKED ISSUE #4895: Option to run PretrainedTransformer in eval mode
While setting `train_parameters` to `False` very often we also may consider disabling dropout/batchnorm, in other words, to run the pretrained model in eval mode.
We've done a little modification to `PretrainedTransformerEmbedder` that allows providing whether the token embedder should be forced to `eval` mode during the training phase.

Do you this feature might be handy? Should I open a PR?

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Fixes #4895 .

Changes proposed in this pull request:

- If `train_parameters` in PretrainedTransformerEmbedder is `False`, the transformer's dropout and batch normalization layers are now set to evaluation mode.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Changed
+- If `train_parameters` in PretrainedTransformerEmbedder is `False`, the transformer's dropout and batch normalization layers are now set to evaluation mode.
```
Comment (epwalsh) [inline CHANGELOG.md:11]: We already have a `### Changed` section on line 13 below, you can just put this bullet point there.

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 2]
```diff
@@ -122,6 +124,7 @@ def __init__(
         self._num_added_tokens = self._num_added_start_tokens + self._num_added_end_tokens
 
         if not train_parameters:
+            self.transformer_model.eval()
             for param in self.transformer_model.parameters():
                 param.requires_grad = False
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: This might not be necessary anymore

**Review by epwalsh — COMMENTED** [review]
Nice job! I just have a couple comments.

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 2 above)
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: Got it. [This Stack Overflow post](https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch) says that it's still best practice to turn off gradients computation. Thoughts?

Comment (epwalsh) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: Reading that, sounds like we may want to use `with torch.no_grad()` in the forward pass?

I'm also curious what happens after this class is initialized as part of a submodule for a `Model`, when `Model.train()` is called. Will that revert the `.eval()` call within this submodule?

Comment (epwalsh) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: Might be worth having a test for that

File: CHANGELOG.md (hunk 1 above)
Comment (ArjunSubramonian) [reply CHANGELOG.md]: Done!

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 2 above)
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:141]: Nice catch! I fixed this issue, albeit in a hacky way (https://stackoverflow.com/questions/61980943/how-can-i-keep-a-pytorch-submodule-in-eval-mode, https://stackoverflow.com/questions/394770/override-a-method-at-instance-level). Let me know what you think. I also added a test case to ensure that PretrainedTrransformerEmbedder's transformer model remains in eval mode even when module that instantiates it is switched to training mode.

Comment (epwalsh) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:141]: One thing you could do is just rename the existing `.forward()` method to `._forward()`, and then implement a new `.foward()` like:

```python
def forward(self, ...):
    if self.train_parameters:
        return self._forward(...)
    else:
        with torch.no_grad():
            return self._foward(...)
```

Comment (epwalsh) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:141]: ... that said, I just realized there are other places in our code-base that would expect `requires_grad` to be `False` when this module is initialized, if we're not training it. So yea, I think we do need to keep these lines.

File: tests/modules/token_embedders/pretrained_transformer_embedder_test.py [hunk 3]
```diff
@@ -330,3 +330,22 @@ def test_embeddings_resize(self):
             enhanced_token_embedder.transformer_model.embeddings.word_embeddings.num_embeddings
             == 28997
         )
+
+    def test_eval_mode(self):
+        token_embedder = PretrainedTransformerEmbedder("bert-base-uncased", train_parameters=False)
```
Comment (epwalsh) [inline tests/modules/token_embedders/pretrained_transformer_embedder_test.py:335]: You could use the "epwalsh/bert-xsmall-dummy" model instead of "bert-base-uncased" to make this test more lightweight.

Comment (epwalsh) [reply tests/modules/token_embedders/pretrained_transformer_embedder_test.py:335]:
```suggestion
        token_embedder = PretrainedTransformerEmbedder("epwalsh/bert-xsmall-dummy", train_parameters=False)
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 4]
```diff
@@ -122,6 +125,18 @@ def __init__(
...
+            # so we have to explicitly set training = False
+            self.transformer_model.training = False
+
+            # Override train in transformer_model to prevent it from changing modes
+            def _train(self, mode):
+                return self
+
+            setattr(
+                self.transformer_model, "train", types.MethodType(_train, self.transformer_model)
+            )
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:138]: This makes me a bit nervous. Another option would be to add something like this to the `.forward()` method below (after line 182):

```python
if self.train_parameters and self.training and self.transformer_model.training:
    self.transformer_model.eval()
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 5]
```diff
@@ -122,6 +125,18 @@ def __init__(
         self._num_added_tokens = self._num_added_start_tokens + self._num_added_end_tokens
 
         if not train_parameters:
+            # Calling transformer_model.eval() won't change anything now,
+            # so we have to explicitly set training = False
```
[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: Because `.eval()` just calls `.train(False)`

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 4 above)
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:138]: I like this approach much better, thanks for the suggestion :)

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py:129]: Calling `.eval()` or `.train(False)` will recursively call the same on all submodules, and also set `training = False` for all submodules (https://github.com/pytorch/pytorch/blob/d490e0120f32dcbb8b23e11eebd638b96b4b0898/torch/nn/modules/module.py#L1594). Isn't that what we want?

[the remark this answers is not in the document]
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py]: Yes, you're right, that is what we want. I just updated the code to be more robust.

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 4 above)
Comment (ArjunSubramonian) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py]: One issue with this though is that inspecting `.training` will not alway produce the expected result.

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 6]
```diff
@@ -122,6 +125,17 @@ def __init__(
...
+            # Override train in transformer_model to prevent it from changing modes
+            def _train(self, mode):
+                self.training = False
+                for module in self.children():
+                    module.train(False)
+                return self
+
+            setattr(
+                self.transformer_model, "train", types.MethodType(_train, self.transformer_model)
+            )
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:137]: Instead of patching the `.train()` method on `self.transformer` model, why not just override the `.train()` method `PretrainedTransformerEmbedder`?

We could implement it like

```python
@overrides
def train(self, mode: bool = True):
    self.training = mode
    for name, module in self.named_children():
        if not self.train_parameters and name == "transformer_model":
            module.eval()
        else:
            module.train(mode)
    return self
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import logging
 import math
 from typing import Optional, Tuple, Dict, Any
+import types
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:4]:
```suggestion
```

**Review by epwalsh — APPROVED** [review]
LGTM! Nice job!

**Comment by nelson-liu** [comment]
I realize I should have spoken up sooner about this, but would it be possible to change this to add another parameter that controls eval vs. non-eval mode, versus overriding the default behavior of `train_parameters` (asking before the next release goes out and it'd become a breaking change)? In particular, this behavior is pretty different from the setting where you use `Elmo` with `requires_grad=False`. In that setting, the Elmo weights actually isn't in eval mode.

There are also settings where you want the default behavior (i.e., non-eval mode, but frozen parameters). In particular, there's been recent work on trying to do parameter-efficient fine-tuning by, for instance, only fine-tuning the bias terms of the transformer (https://nlp.biu.ac.il/~yogo/bitfit.pdf). In this case, it's much more ergonomic to set `train_parameters` to False and then have a regex for `[["^_text_field_embedder.token_embedder_tokens.transformer_model.*bias"], {"requires_grad": true}]`, versus the having to write a regex like `[["^_text_field_embedder.token_embedder_tokens.transformer_model.*(?<!bias)$"], {"requires_grad": false}]`.

Lastly, I think that semantically `train_parameters` doesn't necessarily imply eval mode---they seem like distinct ways of modifying the model (e.g., how the `Elmo` embedder works right now).
